src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,330 @@
+
+exprt symbol_factoryt::allocate_object(
+  code_blockt &assignments,
+  const exprt &target_expr,
+  const typet &allocate_type,
+  const bool static_lifetime)
+{
+  const symbolt &aux_symbol=
+    c_new_tmp_symbol(
+      symbol_table,
+      loc,
+      allocate_type,
+      static_lifetime);
+  symbols_created.push_back(&aux_symbol);
+
+  const typet &allocate_type_resolved=ns.follow(allocate_type);
+  const typet &target_type=ns.follow(target_expr.type().subtype());
+  bool cast_needed=allocate_type_resolved!=target_type;
+
+  exprt aoe=address_of_exprt(aux_symbol.symbol_expr());
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+
+  // Add the following code to assignments:
+  //   <target_expr> = &tmp$<temporary_counter>
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  assignments.add(assign);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  assignments - The code block to add code to
+  expr - The expression which we are generating a non-determinate value for
+
+ Outputs:
+
+ Purpose: Creates a nondet for expr, including calling itself recursively to
+          make appropriate symbols to point to if expr is a pointer.
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  code_blockt &assignments,
+  const exprt &expr)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_blockt non_null_inst;
+
+    exprt allocated=allocate_object(non_null_inst, expr, subtype, false);
+
+    exprt init_expr;
+    if(allocated.id()==ID_address_of)
+    {
+      init_expr=allocated.op0();
+    }
+    else
+    {
+      init_expr=dereference_exprt(allocated, allocated.type().subtype());
+    }
+    gen_nondet_init(non_null_inst, init_expr);
+
+    if(assume_non_null)
+    {
+      // Add the following code to assignments:
+      // <expr> = <aoe>;
+      assignments.append(non_null_inst);
+    }
+    else
+    {
+      // Add the following code to assignments:
+      //           IF !(NONDET(_Bool) == FALSE) THEN GOTO <label1>
+      //           <expr> = <null pointer>
+      //           GOTO <label2>
+      // <label1>: <expr> = &tmp$<temporary_counter>;
+      //           <code from recursive call to gen_nondet_init() with
+      //             tmp$<temporary_counter>>
+      // And the next line is labelled label2
+      auto set_null_inst=code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      const exprt false_expr=from_integer(0, c_bool_typet(1));
+
+      code_ifthenelset null_check;
+      null_check.cond()=
+        equal_exprt(c_get_nondet_bool(c_bool_typet(1)), false_expr);
```

Comment:
To address the equality-without-matching-types issue you should probably just do
```
null_check.cond()=side_effect_expr_nondett(bool_typet());
```
The issue is the `c_bool_typet` placed in here. Apologies for not spotting that any earlier.